includes/transforms.js

```diff
@@ -0,0 +1,90 @@
+// Copyright © 2023 IBM Corp. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+'use strict';
+
+const { Duplex, PassThrough, Transform } = require('node:stream');
+const debug = require('debug');
+
+/**
+ * Input: stream of elements
+ * Output: stream of arrays of batchSize elements each
+ */
+class BatchingStream extends Transform {
+  constructor(batchSize, highWaterMarkScale = 1) {
+    super({ objectMode: true, readableHighWaterMark: highWaterMarkScale, writableHighWaterMark: highWaterMarkScale * batchSize });
+    this.log = debug(('couchbackup:transform:batch'));
+    this.log(`Batching to size ${batchSize} with scale ${highWaterMarkScale}`);
+    this.batchSize = batchSize;
+    this.batch = [];
+    this.batchId = 0;
+  }
+
+  writeBatch(callback) {
+    if (this.batch.length > 0) {
+      this.log(`Writing batch ${this.batchId} with ${this.batch.length} elements.`);
+      this.push(this.batch);
+      this.batch = [];
+      this.batchId++;
+    }
+    callback();
+  }
+
+  _transform(element, encoding, callback) {
+    if (this.batch.push(element) === this.batchSize) {
+      this.writeBatch(callback);
+    } else {
+      callback();
+    }
+  }
+
+  _flush(callback) {
+    this.log('Flushing batch transform.');
+    this.writeBatch(callback);
```

Comment:
A comment here would be recommended as to why this is called for the last batch.

Reply:
added in https://github.com/IBM/couchbackup/pull/643/commits/3890dac2721963a407173466349af3c0d348aa91